ethclient/simulated/backend_test.go

```diff
@@ -0,0 +1,318 @@
+	}
+}
+
+func TestSendTransaction(t *testing.T) {
+	sim := simTestBackend(testAddr)
+	defer sim.Close()
+
+	client := sim.Client()
+	ctx := context.Background()
+
+	signedTx, err := newTx(sim, testKey)
+	if err != nil {
+		t.Errorf("could not create transaction: %v", err)
+	}
+	// send tx to simulated backend
+	err = client.SendTransaction(ctx, signedTx)
+	if err != nil {
+		t.Errorf("could not add tx to pending block: %v", err)
+	}
+	sim.Commit(false)
+	block, err := client.BlockByNumber(ctx, big.NewInt(1))
+	if err != nil {
+		t.Errorf("could not get block at height 1: %v", err)
+	}
+
+	if signedTx.Hash() != block.Transactions()[0].Hash() {
+		t.Errorf("did not commit sent transaction. expected hash %v got hash %v", block.Transactions()[0].Hash(), signedTx.Hash())
+	}
+}
+
+// TestFork check that the chain length after a reorg is correct.
+// Steps:
+//  1. Save the current block which will serve as parent for the fork.
+//  2. Mine n blocks with n ∈ [0, 20].
+//  3. Assert that the chain length is n.
+//  4. Fork by using the parent block as ancestor.
+//  5. Mine n+1 blocks which should trigger a reorg.
+//  6. Assert that the chain length is n+1.
+//     Since Commit() was called 2n+1 times in total,
+//     having a chain length of just n+1 means that a reorg occurred.
+func TestFork(t *testing.T) {
+	t.Parallel()
+	testAddr := crypto.PubkeyToAddress(testKey.PublicKey)
+	sim := simTestBackend(testAddr)
+	defer sim.Close()
+
+	client := sim.Client()
+	ctx := context.Background()
+
+	// 1.
+	parent, _ := client.HeaderByNumber(ctx, nil)
+
+	// 2.
+	n := int(rand.Int31n(21))
+	for i := 0; i < n; i++ {
+		sim.Commit(false)
+	}
+
+	// 3.
+	b, _ := client.BlockNumber(ctx)
+	if b != uint64(n) {
+		t.Error("wrong chain length")
+	}
+
+	// 4.
+	sim.Fork(parent.Hash())
+
+	// 5.
+	for i := 0; i < n+1; i++ {
+		sim.Commit(false)
+	}
+
+	// 6.
+	b, _ = client.BlockNumber(ctx)
+	if b != uint64(n+1) {
+		t.Error("wrong chain length")
+	}
+}
+
+// TestForkResendTx checks that re-sending a TX after a fork
+// is possible and does not cause a "nonce mismatch" panic.
+// Steps:
+//  1. Save the current block which will serve as parent for the fork.
+//  2. Send a transaction.
+//  3. Check that the TX is included in block 1.
+//  4. Fork by using the parent block as ancestor.
+//  5. Mine a block, Re-send the transaction and mine another one.
+//  6. Check that the TX is now included in block 2.
+func TestForkResendTx(t *testing.T) {
+	t.Parallel()
+	testAddr := crypto.PubkeyToAddress(testKey.PublicKey)
+	sim := simTestBackend(testAddr)
+	defer sim.Close()
+
+	client := sim.Client()
+	ctx := context.Background()
+
+	// 1.
+	parent, _ := client.HeaderByNumber(ctx, nil)
+
+	// 2.
+	tx, err := newTx(sim, testKey)
+	if err != nil {
+		t.Fatalf("could not create transaction: %v", err)
+	}
+	client.SendTransaction(ctx, tx)
+	sim.Commit(false)
+
+	// 3.
+	// Note this test is revised from upstream since we cannot get the receipt
```

Comment:
can we use sim.Commit(true) and then test the receipts if possible? I don't know if accepting a blk here would go against the whole testing logic though.

Reply:
unfortunately that breaks the test:
--- FAIL: TestForkResendTx (0.00s)
    backend_test.go:227: forking: unable to invoke writeKnownBlock: cannot orphan finalized block at height: 1 to common block at height: 0